### PR TITLE
feat: support win32

### PR DESCRIPTION
**Pull request by waitingsong**
it should stop with --port if start cluster with custom port number

##### Checklist

- [x] `npm test` passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows commit guidelines

### CODE REVIEW

**Comment by fengmk2** [comment]
@atian25 add appveyor tests

**Comment by atian25** [comment]
but `egg-scripts start` is not support windows yet, so `appveyor` will break, should we still open it?

**Comment by waitingsong** [comment]
I run test local all passed under win7. not support ?

**Comment by atian25** [comment]
https://github.com/eggjs/egg-scripts/blob/master/lib/cmd/start.js#L180-L196

`tails` is support at win?  maybe you had install some tool such as git bin

**Comment by waitingsong** [comment]
oh, git installed

**Comment by waitingsong** [comment]
```js
 stdout = stdout.split(/\n+/g).splice(0, 100);
 +            this.logger.error(stdout);
 +            this.logger.error('Start failed, see %s', stderr);
```
need convert stdout to string by  stdout.join('\n') ?

File: lib/cmd/start.js [hunk 1]
```diff
@@ -185,11 +185,22 @@ class StartCommand extends Command {
...
-          break;
+          if (process.platform !== 'win32') {
+            const [ stdout ] = yield exec('tail -n 100 ' + stderr);
+            this.logger.error(stdout);
+            this.logger.error('Start failed, see %s', stderr);
+            isSuccess = false;
+            break;
+          } else {
+            const str = fs.readFileSync(stderr, 'utf-8');
+            let stdout = str ? str.slice(0, 50000).replace(/\r/g, '\n') : '';
```
Comment (popomore) [inline lib/cmd/start.js]: you can simply print all on windows

File: lib/cmd/stop.js [hunk 2]
```diff
@@ -17,13 +17,8 @@ class StopCommand extends Command {
...
 
   * run(context) {
-    /* istanbul ignore next */
-    if (process.platform === 'win32') {
-      this.logger.warn('Windows is not supported, try to kill master process which command contains `start-cluster` or `--type=egg-server` yourself, good luck.');
-      process.exit(0);
-    }
-
     const { argv } = context;
+    const port = argv.port || (process.env && process.env.PORT);
```
Comment (popomore) [inline lib/cmd/stop.js]: Is there other way to stop without port

Comment (waitingsong) [reply lib/cmd/stop.js]: 有其他方法。但相当复杂
1、修改egg-cluster, 在启动worker进程时，更新窗口的title 
2、在关闭服务时借助node-ffi 找到任何一个匹配title的worker窗口获取其pid
3、借助node-ffi 调用 ntdll.NtQueryInformationProcess() 方法，根据pid找到父级pid（即master进程），然后kill。

1、2、我计划实现，并且可根据参数隐藏worker的窗口。
因为master进程没有窗口，也就无法设置其窗口title，所以只有在worke上面设置了。

File: lib/cmd/start.js (hunk 1 above)
Comment (waitingsong) [reply lib/cmd/start.js]: 我是参考源代码 tail 前100行，我以一行500字节*100行计算来截取 50000字节。

File: lib/cmd/stop.js (hunk 2 above)
Comment (popomore) [reply lib/cmd/stop.js]: Or we can store port when app starts, so it’s unnecessary to give the port argument.

Comment (waitingsong) [reply lib/cmd/stop.js]: 能记录所有进程 pid 那当然就方便

**Comment by atian25** [comment]
so just like Nginx `run/pid`? @popomore 

发自我的 iPhone

**Comment by atian25** [comment]
https://ci.appveyor.com/project/eggjs/egg-scripts

had enabled, need an empty commit to trigger it.

**Comment by waitingsong** [comment]
commit on PR, but seems not triggered. only master branch?

**Comment by atian25** [comment]
try again and see, just delete and recreate it.

File: lib/cmd/start.js [hunk 3]
```diff
@@ -185,11 +185,22 @@ class StartCommand extends Command {
...
-          isSuccess = false;
-          break;
+          if (process.platform !== 'win32') {
+            const [ stdout ] = yield exec('tail -n 100 ' + stderr);
+            this.logger.error(stdout);
+            this.logger.error('Start failed, see %s', stderr);
+            isSuccess = false;
+            break;
+          } else {
+            const str = fs.readFileSync(stderr, 'utf-8');
```
Comment (atian25) [inline lib/cmd/start.js]: use `yield fs.readFile`

**Pull request closed without merging**